codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -191,16 +187,61 @@ public String path() {
         return path;
     }
 
+    @UnstableApi
+    final boolean hasDecodedParameters() {
+        return params != null;
+    }
+
     /**
      * Returns the decoded key-value parameter pairs of the URI.
      */
     public Map<String, List<String>> parameters() {
         if (params == null) {
-            params = decodeParams(uri, pathEndIdx(), charset, maxParams, semicolonIsNormalChar);
+            params = new LinkedHashMap<String, List<String>>();
+            decodeParams(uri, pathEndIdx(), charset, MAP_PARAMETER_COLLECTOR, params, maxParams, semicolonIsNormalChar);
         }
         return params;
     }
 
+    /**
+     * Decodes {@link #uri()} parameters, reporting them in the provided {@code collector}'s map.
+     */
+    public void decodeParameters(Map<String, List<String>> parameters) {
```

Comment:
I think renaming the param makes things more clear 

```suggestion
    public void decodeParameters(Map<String, List<String>> out) {
```